### PR TITLE
planner, executor: enable inline projection for merge join

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve?

Problem Summary:

part of #14428

### What is changed and how it works?

What's Changed:

enable inline projection for merge join

How it Works:

`markChildrenUsedCols`is applied to joiner.

### Check List 

Tests 

- Unit test

Side effects

### Release note 

----

### benchmark based on [BenchmarkMergeJoinExec](https://github.com/pingcap/tidb/blob/bbac2b2/executor/benchmark_test.go#L1507)

two data sources with columns of type {longlong, double, varstring(5*2014)}

### case 1
no modify:
```
BenchmarkMergeJoinExec/merge_join_(outerRows:300000,_innerRows:300000,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-4         	       1	2407966792 ns/op	58474936 B/op	    4070 allocs/op
BenchmarkMergeJoinExec/merge_join_(outerRows:3000,_innerRows:300000,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-4           	       2	 514012846 ns/op	58471576 B/op	    4053 allocs/op
BenchmarkMergeJoinExec/merge_join_(outerRows:30,_innerRows:300000,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-4             	       2	 709402071 ns/op	62614336 B/op	    4007 allocs/op
BenchmarkMergeJoinExec/merge_join_(outerRows:300000,_innerRows:330000,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-4         	       1	2381562052 ns/op	58466440 B/op	    4064 allocs/op
```

### case 2
childrenUsedCols (and output schema) changed to: 
```
{true, false, false},
{false, false, false},
```

significant improvement due to less column copy

```
BenchmarkMergeJoinExec/merge_join_(outerRows:300000,_innerRows:300000,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-4         	      10	 104559070 ns/op	  717604 B/op	    3940 allocs/op
BenchmarkMergeJoinExec/merge_join_(outerRows:3000,_innerRows:300000,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-4           	      67	  18877980 ns/op	  727133 B/op	    3928 allocs/op
BenchmarkMergeJoinExec/merge_join_(outerRows:30,_innerRows:300000,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-4             	      54	  23297968 ns/op	 4873998 B/op	    3884 allocs/op
BenchmarkMergeJoinExec/merge_join_(outerRows:300000,_innerRows:330000,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-4         	      10	 104550756 ns/op	  722180 B/op	    3940 allocs/op
```

### case 3
childrenUsedCols (and output schema) changed to: 
```
{true, false, false},
{false, true, false},
```

significant improvement, slightly slower than case 2, which is reasonable because output schema has an extra `double` column

```
BenchmarkMergeJoinExec/merge_join_(outerRows:300000,_innerRows:300000,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-4         	      10	 107155546 ns/op	  759140 B/op	    3957 allocs/op
BenchmarkMergeJoinExec/merge_join_(outerRows:3000,_innerRows:300000,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-4           	      54	  22822921 ns/op	  761917 B/op	    3943 allocs/op
BenchmarkMergeJoinExec/merge_join_(outerRows:30,_innerRows:300000,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-4             	      37	  28515056 ns/op	 4910528 B/op	    3899 allocs/op
BenchmarkMergeJoinExec/merge_join_(outerRows:300000,_innerRows:330000,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)-4         	      10	 109264842 ns/op	  757220 B/op	    3954 allocs/op
```

### CODE REVIEW

File: planner/core/task.go [hunk 1]
```diff
@@ -574,7 +574,6 @@ func (p *PhysicalMergeJoin) attach2Task(tasks ...task) task {
 	lTask := finishCopTask(p.ctx, tasks[0].copy())
 	rTask := finishCopTask(p.ctx, tasks[1].copy())
 	p.SetChildren(lTask.plan(), rTask.plan())
-	p.schema = BuildPhysicalJoinSchema(p.JoinType, p)
```
Comment (lance6716) [inline planner/core/task.go]: TODO: I didn't know why we need this

File: planner/cascades/testdata/integration_suite_out.json [hunk 2]
```diff
@@ -1079,6 +1079,19 @@
         "Result": [
           "1 1"
         ]
+      },
+      {
+        "SQL": "select /*+ MERGE_JOIN(t1) */ t1.b, t2.b from t1 inner join t2 on t1.a = t2.a;",
+        "Plan": [
+          "HashRightJoin_9 12500.00 root inner join, equal:[eq(test.t1.a, test.t2.a)]",
```
Comment (lance6716) [inline planner/cascades/testdata/integration_suite_out.json]: It's still HashRightJoin 🤔

Comment (SunRunAway) [reply planner/cascades/testdata/integration_suite_out.json]: cascades planner currently does not support hint. It's ok here.

File: planner/core/task.go (hunk 1 above)
Comment (SunRunAway) [reply planner/core/task.go]: Because we have already pruned extra columns in logical optimization here, https://github.com/pingcap/tidb/blob/d1aa2f5b09d5763c53838310318252b574de154e/planner/core/rule_column_pruning.go#L327

This line rebuilds the schema of Join and reverts the effect of what logical optimization did before. So it should be deleted.

Comment (SunRunAway) [reply planner/core/task.go]: And this [line](https://github.com/pingcap/tidb/blob/6d00e2169bda543797aa3136cfcb5332911cd64d/planner/implementation/join.go#L67) in cascades planner should be deleted either, for the same reason

File: executor/benchmark_test.go [hunk 3]
```diff
@@ -225,6 +225,7 @@ func buildMockDataSourceWithIndex(opt mockDataSourceParameters, index []int) *mo
 
 type aggTestCase struct {
 	// The test table's schema is fixed (aggCol Double, groupBy LongLong).
+
```
Comment (SunRunAway) [inline executor/benchmark_test.go]: This change is not necessary.

File: executor/benchmark_test.go [hunk 4]
```diff
@@ -500,6 +501,7 @@ func buildWindowExecutor(ctx sessionctx.Context, windowFunc string, funcs int, f
 
 type windowTestCase struct {
 	// The test table's schema is fixed (col Double, partitionBy LongLong, rawData VarString(16), col LongLong).
+
```
Comment (SunRunAway) [inline executor/benchmark_test.go]: This change is not necessary.

File: executor/executor_required_rows_test.go [hunk 5]
```diff
@@ -828,6 +828,10 @@ func (mp *mockPlan) GetExecutor() Executor {
 	return mp.exec
 }
 
+func (mp *mockPlan) Schema() *expression.Schema {
```
Comment (SunRunAway) [inline executor/executor_required_rows_test.go]: Is this change necessary?

File: executor/benchmark_test.go (hunk 3 above)
Comment (lance6716) [reply executor/benchmark_test.go]: `goword` assumed line 227 was a comment for `windowFunc`, and not start with word `windowFunc`. So I added a blank line.

File: executor/benchmark_test.go (hunk 4 above)
Comment (lance6716) [reply executor/benchmark_test.go]: ditto

File: executor/executor_required_rows_test.go (hunk 5 above)
Comment (lance6716) [reply executor/executor_required_rows_test.go]: https://github.com/pingcap/tidb/blob/b7894dfefbe4a435ca709597eadd72fb7a774f6a/executor/executor_required_rows_test.go#L808

here using `mockPlan` as children of a `MergeJoinPlan`, thus when building need to implement `*mockPlan.Schema()`

File: executor/benchmark_test.go (hunk 3 above)
Comment (SunRunAway) [reply executor/benchmark_test.go]: Hi, I think it's better to move this comment above `type aggTestCase`

```go
// The test table's schema is fixed (aggCol Double, groupBy LongLong).
type aggTestCase struct {
}
```

**Comment by lance6716** [comment]
friendly ping @SunRunAway

**Comment by SunRunAway** [comment]
> friendly ping @SunRunAway

Sorry for the late reply. It's a great job. Do you have any benchmark data? You can put the benchmark
results right in the pull request description

**Comment by SunRunAway** [comment]
Hi, @lance6716
You can take a look at [`BenchmarkMergeJoinExec`](https://github.com/pingcap/tidb/blob/bbac2b2/executor/benchmark_test.go#L1507) and try to adjust to fit inline projection situation. This [link](https://github.com/pingcap/community/blob/master/contributors/README.md#benchmark
) may help you.

**Comment by lance6716** [comment]
> Hi, @lance6716
> You can take a look at [`BenchmarkMergeJoinExec`](https://github.com/pingcap/tidb/blob/bbac2b2/executor/benchmark_test.go#L1507) and try to adjust to fit inline projection situation. This [link](https://github.com/pingcap/community/blob/master/contributors/README.md#benchmark) may help you.

Done

**Comment by SunRunAway** [comment]
> > Hi, @lance6716
> > You can take a look at [`BenchmarkMergeJoinExec`](https://github.com/pingcap/tidb/blob/bbac2b2/executor/benchmark_test.go#L1507) and try to adjust to fit inline projection situation. This [link](https://github.com/pingcap/community/blob/master/contributors/README.md#benchmark) may help you.
> 
> Done

Hi, @lance6716
Well done.
Case 3 could persist in `BenchmarkMergeJoinExec` as a sub-bench.

**Comment by lance6716** [comment]
> > > Hi, @lance6716
> > > You can take a look at [`BenchmarkMergeJoinExec`](https://github.com/pingcap/tidb/blob/bbac2b2/executor/benchmark_test.go#L1507) and try to adjust to fit inline projection situation. This [link](https://github.com/pingcap/community/blob/master/contributors/README.md#benchmark) may help you.
> > 
> > 
> > Done
> 
> Hi, @lance6716
> Well done.
> Case 3 could persist in `BenchmarkMergeJoinExec` as a sub-bench.

```
goos: darwin
goarch: amd64
pkg: github.com/pingcap/tidb/executor
BenchmarkMergeJoinExec/merge_join_(outerRows:300000,_innerRows:300000,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)_InlineProj:false-4         	       1	2449115485 ns/op	58474936 B/op	    4070 allocs/op
BenchmarkMergeJoinExec/merge_join_(outerRows:300000,_innerRows:300000,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)_InlineProj:true-4          	      10	 108524066 ns/op	  759140 B/op	    3957 allocs/op
BenchmarkMergeJoinExec/merge_join_(outerRows:3000,_innerRows:300000,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)_InlineProj:false-4           	       2	 680593549 ns/op	58471576 B/op	    4053 allocs/op
BenchmarkMergeJoinExec/merge_join_(outerRows:3000,_innerRows:300000,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)_InlineProj:true-4            	      51	  22117074 ns/op	  762900 B/op	    3943 allocs/op
BenchmarkMergeJoinExec/merge_join_(outerRows:30,_innerRows:300000,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)_InlineProj:false-4             	       3	 509314305 ns/op	62615528 B/op	    4005 allocs/op
BenchmarkMergeJoinExec/merge_join_(outerRows:30,_innerRows:300000,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)_InlineProj:true-4              	      43	  28141019 ns/op	 4911220 B/op	    3900 allocs/op
BenchmarkMergeJoinExec/merge_join_(outerRows:300000,_innerRows:330000,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)_InlineProj:false-4         	       1	1472482612 ns/op	58485000 B/op	    4072 allocs/op
BenchmarkMergeJoinExec/merge_join_(outerRows:300000,_innerRows:330000,_concurency:4,_outerJoinKeyIdx:_[0_1],_innerJoinKeyIdx:_[0_1],_NeedOuterSort:false)_InlineProj:true-4          	      10	 108598788 ns/op	  758148 B/op	    3955 allocs/op
```

**Comment by Reminiscent** [comment]
@lance6716 Thanks for your contribution. Please resolve the conflicts. Thanks!

**Comment by lance6716** [comment]
@Reminiscent conflict resolved, waiting for a approval 😄

**Comment by lance6716** [comment]
ping @SunRunAway @Reminiscent
